### ISSUE TITLE
Odd InvalidData error

### ISSUE THREAD

**Issue by mt-caret** [simd-lite/simd-json#147]
Hi, I've been using simd-json as a drop-in replacement for serde_json, and was attempting to speed up deserialization of a very large (45GB) JSON file of an object containing an array with around a million elements. However, it seems like simd_json throws an error (`Error: Custom { kind: InvalidData, error: Error { index: 9, character: 'o', error: ExpectedObjectContent } }`) while serde_json can parse the file just fine. I've tried to pinpoint the location where the error occurs with binary search, and I've whittled down the JSON file which reproduces the error down to 5.6GB. [The compressed version of the file is here](https://drive.google.com/file/d/1SE4xB2qlJjO43Bj8lfWTXHG6czToMrzZ/view?usp=sharing).

I've created a [minimal example](https://github.com/mt-caret/simd_json_test) which deserializes to a serde_json::Value and reproduces the error. Using simd-json:
```
$ cargo run --release ../2020-08-12.subset4.json
    Finished release [optimized] target(s) in 0.04s
     Running `target/release/simd_json_test ../2020-08-12.subset4.json`
Error: Custom { kind: InvalidData, error: Error { index: 9, character: 'o', error: ExpectedObjectContent } }
```

Using serde_json:
```
$ cargo run --release -- --disable-simd ../2020-08-12.subset4.json
    Finished release [optimized] target(s) in 0.04s
     Running `target/release/simd_json_test --disable-simd ../2020-08-12.subset4.json
```

Am I doing anything wrong here, or is serde_json excessively permissive and the file I'm giving to simd-json somehow malformed?

**Comment by mt-caret** [comment]
So that's why splitting the 5.6GB file in half prevents that error from occurring, thanks for the explanation!

> The easiest solution would probably be splitting files to < 4GB but I suspect that might not be possible?

Unfortunately, it's probably not possible for my use case.

I'm guessing there's a reason (performance, etc.) why the indices are 32bits, so changing them to 64bits is not really reasonable, right? How much work do you think would be hiding 64bit indices behind a feature flag be? If it's not that much and you're ok with it, I would love to contribute.

**Comment by lemire** [comment]
@mt-caret 

Changing the indices so that they are 64-bit instead of 32-bit is not hard. However, if performance is an important concern, when loading up 5.6GB of data into a memory structure and then accessing it, one should keep in mind that it is a performance anti-pattern. That is, it is a bad design generally speaking...

To put it differently, if you are not using huge pages, you are going to chasing memory pages all the way when loading, and walking them back as you access them. It is going to end up being a memory subsystem benchmark.

The file takes the following form...

```json
{"total_rows":1359244,"offset":0,"rows":[
{"id":"-","key":},
{"id":"-lidonghui","key":},
]}
```

What one almost surely want is a fast iterator on the content of "rows", returning objects one by one or in blocks.

Your origin post refers to a 45GB file. Loading up a 45 GB JSON file in memory all at once is what I would describe as a "smell": it is an indication that something is probably bad in the software design.

You almost always want paging when loading massive files.


@Licenser Are you missing a check when you receive the data? You probably don't want to fail with "ExpectedObjectContent".

cc @jkeiser

**Comment by mt-caret** [comment]
@lemire 

> Changing the indices so that they are 64-bit instead of 32-bit is not hard. However, if performance is an important concern, when loading up 5.6GB of data into a memory structure and then accessing it, one should keep in mind that it is a performance anti-pattern.
> To put it differently, if you are not using huge pages, you are going to chasing memory pages all the way when loading, and walking them back as you access them. It is going to end up being a memory subsystem benchmark.

To sidestep this issue, I'm currently using serde_json instead of simd-json. Judging by how fast the two libraries ingest the JSON documents in question (disregarding the error simd-json emits halfway), the bottleneck still seems to be JSON parsing instead of paging. I guess I'll take a look at the code to take a crack at making the indices larger.

> Your origin post refers to a 45GB file. Loading up a 45 GB JSON file in memory all at once is what I would describe as a "smell": it is an indication that something is probably bad in the software design.

I agree. Unfortunately, working on the 45GB JSON file wasn't really a deliberate choice of mine :cry:
I do extract only what I need and reserialize it with bincode, which brings it down to 20GB or so.

> What one almost surely want is a fast iterator on the content of "rows", returning objects one by one or in blocks.

Is this possible? This was originally what I was trying to do, but I wasn't sure of how to do it. serde_json's StreamDeserializer seemed to only accept a top-level array or objects delimited with newlines, but no support for extracting an array from an object, then streaming it.

**Comment by dtolnay** [comment]
If https://github.com/mt-caret/simd_json_test/blob/master/src/main.rs is representative of how you are using serde_json, you're gonna want to replace it with a [memmap](https://crates.io/crates/memmap) + from_slice.

```rust
let f = File::open(opt.input.clone())?;
let mmap = unsafe { MmapOptions::new().map(&f)? };
let _: Value = serde_json::from_slice(&mmap)?;
```

This drops your demo file from 95 seconds to 38 seconds on my machine.

After that you'll want to replace `Value` with a real data structure that contains only the things you need (*"I do extract only what I need"*). For example if what you need were as follows, deserializing to this `Rows` struct is reduced from 38 seconds to 7 seconds.

```rust
#[derive(Deserialize)]
struct Rows {
    total_rows: usize,
    offset: usize,
    rows: Vec<Row>,
}

#[derive(Deserialize)]
struct Row {
    doc: Doc,
}

#[derive(Deserialize)]
struct Doc {
    versions: BTreeMap<String, Version>,
}

#[derive(Deserialize)]
struct Version {
    description: Option<String>,
}
```

After that, you should look into directly streaming bincode output without holding rows in memory. Something like:

<details>
<summary>main.rs</summary>

```rust
// [dependencies]
// bincode = "1.3"
// memmap = "0.7"
// serde = { version = "1.0", features = ["derive"] }
// serde_json = "1.0"

use memmap::MmapOptions;
use serde::de::{self, DeserializeSeed, IgnoredAny, MapAccess, SeqAccess, Visitor};
use serde::{Deserialize, Deserializer, Serialize};
use std::collections::BTreeMap;
use std::fmt;
use std::fs::File;
use std::io::{self, BufWriter};

struct Stream(bincode::Serializer<BufWriter<File>, bincode::DefaultOptions>);

impl<'de> Visitor<'de> for Stream {
    type Value = ();

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("JSON map containing `total_rows`, `offset`, `rows`")
    }

    fn visit_map<A>(mut self, mut map: A) -> Result<(), A::Error>
    where
        A: MapAccess<'de>,
    {
        #[derive(Deserialize)]
        #[serde(field_identifier, rename_all = "snake_case")]
        enum Key {
            TotalRows,
            Offset,
            Rows,
        }

        while let Some(k) = map.next_key()? {
            match k {
                Key::Rows => map.next_value_seed(Rows(&mut self.0))?,
                _ => drop(map.next_value::<IgnoredAny>()?),
            }
        }

        Ok(())
    }
}

struct Rows<'a>(&'a mut bincode::Serializer<BufWriter<File>, bincode::DefaultOptions>);

impl<'de> DeserializeSeed<'de> for Rows<'_> {
    type Value = ();

    fn deserialize<D>(self, deserializer: D) -> Result<(), D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_seq(self)
    }
}

#[derive(Deserialize, Serialize)]
struct Row {
    doc: Doc,
}

#[derive(Deserialize, Serialize)]
struct Doc {
    versions: BTreeMap<String, Version>,
}

#[derive(Deserialize, Serialize)]
struct Version {
    description: Option<String>,
}

impl<'de> Visitor<'de> for Rows<'_> {
    type Value = ();

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("JSON array of rows")
    }

    fn visit_seq<A>(self, mut seq: A) -> Result<(), A::Error>
    where
        A: SeqAccess<'de>,
    {
        while let Some(row) = seq.next_element::<Row>()? {
            row.serialize(&mut *self.0).map_err(de::Error::custom)?;
        }
        Ok(())
    }
}

fn main() -> io::Result<()> {
    let input = File::open("2020-08-12.subset4.json")?;
    let mmap = unsafe { MmapOptions::new().map(&input)? };
    let mut de = serde_json::Deserializer::from_slice(&mmap);

    let output = File::create("out.bincode")?;
    let writer = BufWriter::new(output);
    let ser = bincode::Serializer::new(writer, bincode::DefaultOptions::new());

    de.deserialize_struct("Stream", &[], Stream(ser))?;
    Ok(())
}
```
</details>

**Comment by lemire** [comment]
I like that I don't need to answer anymore. Thank you @dtolnay.

**Comment by Licenser** [comment]
> @Licenser Are you missing a check when you receive the data? You probably don't want to fail with "ExpectedObjectContent".

Yes that's a oversight :) needs to be fixed! And big +1 to @dtolnay serde shines here since it allows you to visit only certain parts of the binary instead of analyzing it completely, especially since you probably still want to parse/emit row 1-1000000 if row 10000001 has a issue :D

**Comment by Licenser** [comment]
fixed in `v0.3.20`
